plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/conditional/RelationalExpression.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor.conditional;
+
+import org.elasticsearch.ingest.IngestDocument;
+
+public class RelationalExpression implements ConditionalExpression {
+    private final RelationalOperator operator;
+    private final String field;
+    private final Object operand;
+
+    public RelationalExpression(String field, String operatorName, Object operand) {
+        this(field, RelationalOperator.fromString(operatorName), operand);
+    }
+
+    public RelationalExpression(String field, RelationalOperator operator, Object operand) {
+        this.field = field;
+        this.operator = operator;
+        this.operand = operand;
+    }
+
+    public RelationalOperator getOperator() {
+        return operator;
+    }
+
+    public Object getOperand() {
+        return operand;
+    }
+
+    public boolean matches(IngestDocument ingestDocument) {
+        Object other = ingestDocument.getPropertyValue(field, Object.class);
+        if (operator.equals(RelationalOperator.EQUALS)) {
+            if (operand == null) {
+                return other == null;
+            }
+            return operand.equals(other);
+        } else if (operator.equals(RelationalOperator.NOT_EQUALS)) {
+            if (operand == null) {
+                return other != null;
+            }
+            return !operand.equals(other);
+        }
+
+
+        // other operators require comparison between numbers.
+        // comparisons cannot be made with null values.
+        if (operand != null && other != null
```

Comment:
I think we can simplify this and support other number types too. All concrete number classes implement Comparable (unfortunately Number class doesn't). So instead of converting to a primitive number we can convert to Comparable class and compare on that: This would look something like this:

```java
if((operand instanceof Comparable && other instanceof Comparable) == false) {
    throw new IllegalArgumentException("Number isn't comparable");
}

Comparable left = ((Comparable) other);
Comparable right = ((Comparable) operand);
@SuppressWarnings("unchecked")
int cmp = left.compareTo(right);
switch (operator) {
    case LESS_THAN:
        return cmp < 0;
    case LESS_THAN_OR_EQUALS:
        return cmp <= 0;
    case GREATER_THAN:
        return cmp > 0;
    case GREATER_THAN_OR_EQUALS:
        return cmp >= 0;
}
```